File: data_structures/Diameter_Of_Binary_Tree/Diameter_Of_Binary_Tree.py
'''
Given the root of a binary tree, return the length of the diameter of the tree.
The diameter of a binary tree is the length of the longest path between any two nodes in a tree. This path may or may not pass through the root.
The length of a path between two nodes is represented by the number of edges between them.
'''
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: TreeNode) -> int:
        def diameterTreeUtil(root):
            if root == None: return [0,0]
            
            leftTree = diameterTreeUtil(root.left)
            rightTree = diameterTreeUtil(root.right)
            
            internalPath = max(leftTree[1],rightTree[1])
            if leftTree[0] + rightTree[0] + 1 > internalPath: internalPath = leftTree[0] + rightTree[0] + 1
            
            return [max(leftTree[0],rightTree[0])+1,internalPath]
        
        if root == None: return 0
        result = diameterTreeUtil(root)
        return max(result)-1

File: data_structures/Diameter_Of_Binary_Tree/Diameter_Of_Binary_Tree.cpp
#include <bits/stdc++.h>
using namespace std;
struct node
{
	int data;
	struct node *left, *right;
};
struct node *newNode(int data);

int ans=0;
int find(struct node* root)
{  
    if(!root) return 0;

     int l = find(root->left);
     int r = find(root->right);

     ans=max(ans,l+r+1);

     return max(l,r)+1;
}
int diameter(struct node *tree)
{       
	ans=0;
	if(!root) 
          return 0;
        find(root);
          return ans;
}
struct node *newNode(int data)
{
	struct node *node = (struct node *)malloc(sizeof(struct node));
	node->data = data;
	node->left = NULL;
	node->right = NULL;

	return (node);
}

int main()
{

	/* Constructed binary tree is
			 1
			/ \
		   2   3
		  /     \
		 4	     5
	*/
	struct node *root = newNode(1);
	root->left = newNode(2);
	root->right = newNode(3);
	root->left->left = newNode(4);
	root->left->right = newNode(5);

	cout << "Diameter of the given binary tree is " << diameter(root);

	return 0;
}
